src/concrete/ml/quantization/quantized_module_passes.py

```diff
@@ -0,0 +1,300 @@
+
+        Returns:
+            result (PatternDict): a dictionary containing for each Conv/Gemm node for which
+                round PBS can be applied based on power-of-two scaling factors
+        """
+
+        # The Pattern can be described as follows
+        # x = Quant(x) -> stored separately in the second member of the tuple in PatternDict
+        # .... the following ops are stored in the List of the PatternDict
+        # y = Gemm(x, w, b), with w, b produced by a Brevitas quant node
+        #     +---> This is the node for which roundPBS can be adjusted
+        # y = Relu(y)
+        # y = Quant(y)
+        # z = Gemm(y, w2, b2)  -> the output node of the pattern
+
+        self._num_ignored_valid_patterns = 0
+
+        predecessors = self.compute_op_predecessors()
+
+        valid_paths = self.detect_patterns(predecessors)
+
+        valid_paths = self.process_patterns(valid_paths)
+
+        return valid_paths
+
+    def compute_op_predecessors(self) -> PredecessorsType:
+        """Compute the predecessors for each QuantizedOp in a QuantizedModule.
+
+        Stores, for each quantized op, a list of quantized ops that produce its
+        inputs. Currently only the first input of the operations is considered
+        as it is, usually, the encrypted input.
+
+        Returns:
+            result (PredecessorsType): a dictionary containing a hierarchy of op
+                predecessors
+        """
+
+        # Initialize the list of predecessors with tensors that are graph inputs
+        predecessors: PredecessorsType = defaultdict(list)
+
+        for (node_inputs, node_op) in self._qmodule.quant_layers_dict.values():
+            # The first input node contains the encrypted data
+            enc_input_node = node_inputs[0]
+
+            assert_true(
+                enc_input_node in self._qmodule.quant_layers_dict
+                or enc_input_node in self._qmodule.ordered_module_input_names
+            )
+            pred = self._qmodule.quant_layers_dict.get(enc_input_node, (None, None))
+            # Get the quantized op that produces the current op's input
+            pred_with_output = (pred[1], enc_input_node)
+            predecessors[node_op].append(pred_with_output)
+        return predecessors
+
+    def match_path_pattern(
+        self,
+        predecessors: PredecessorsType,
+        nodes_in_path: List[Optional[QuantizedOp]],
+        input_producer_of_path: Optional[QuantizedOp],
+    ) -> bool:
+        """Determine if a pattern has the structure that makes it viable for roundPBS.
+
+        Args:
+            predecessors (PredecessorsType): Module predecessor operation list
+            nodes_in_path (List[QuantizedOp]): list of quantized ops in the pattern
+            input_producer_of_path (Optional[QuantizedOp]): operation that produces the input
+
+        Returns:
+            result (bool): whether the pattern can be optimized
+        """
+
+        if len(nodes_in_path) != 3:
```

Comment:
all of this method's steps need at least a comment explaining why we return these values as it's not very clear